### PR TITLE
common: Fix large requests with TLS

### PR DESCRIPTION
**Pull request by martinpitt**
(1) With a GTlsServerConnection, the GSource callback is not called
again if there is still pending data in GnuTLS'es buffer [1]. I. e. if
the input buffer had more than 4 KiB of pending data,
`on_request_input()` previously only read the first 4 KiB of data, and
then never tried again as there was no further callback.

(2) Also, there are legitimate HTTP requests (observed in the wild with
Kerberos SSO logins) which are slightly bigger than 4 KiB, so
considering that as a maximum request size is not enough.

Incidentally, larger requests up to 8 KiB happen to work because the
support for request pipelining leaves buffer space for up to 2 requests,
and it really only tests the buffer size instead of the individual
request size. But combined with the Gio bug above, they only worked for
plain HTTP, not HTTPS.

(1) and (2) together caused large TLS requests to time out.

As a minimally intrusive first fix, double the maximum request size to
fix (2), and make sure we read a full maximum size block to work around
(1). This will still hang for several large requests that are pipelined;
for these this needs to be changed into a loop. This will be done in a
follow-up. As a first mitigation, add one extra byte so that
parse_and_process_request() rejects requests that are > 8 KiB, instead
of hanging.

Use the `cockpit_webserver_request_maximum` constant instead of a
repeated hardcoded number. This constant isn't used anywhere else, so
drop the declaration in the header and make it an actual constant.

https://bugzilla.redhat.com/show_bug.cgi?id=1785509

[1] https://gitlab.gnome.org/GNOME/glib-networking/issues/20

### CODE REVIEW

**Comment by martinpitt** [comment]
Unfortunately I was unable to reproduce the original bug in the unit tests, so they only serve as regression test. Manual reproducer:
```
mkdir /tmp/x
XDG_CONFIG_DIRS=/tmp/x ./remotectl certificate --ensure
G_MESSAGES_DEBUG=all ./cockpit-ws -p 9999 --no-tls
```
and in a second terminal:
```
curl -k -H "Foo: bar" -H "Authorization: Negotiate $(printf '%0.5000i' 1)" https://localhost:9999/cockpit/login
```

**Comment by martinpitt** [comment]
I added an integration test for this now. On current master it fails like this:
```
+ curl -s -S -k -H "Authorization: Negotiate $(printf '%0.5000i' 1)" https://localhost:9000/
curl: (52) Empty reply from server
Traceback (most recent call last):
  File "/usr/lib64/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib64/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "test/verify/check-connection", line 454, in testCommandline
    self.assertIn('A Custom Title', m.execute('''curl -s -S -k -H "Authorization: Negotiate $(printf '%0.5000i' 1)" https://localhost:9000/'''))
  File "/var/home/martin/upstream/bots/machine/machine_core/ssh_connection.py", line 344, in execute
    raise subprocess.CalledProcessError(proc.returncode, command, output=output)
subprocess.CalledProcessError: Command 'curl -s -S -k -H "Authorization: Negotiate $(printf '%0.5000i' 1)" https://localhost:9000/' returned non-zero exit status 52.
```

Note that this fix is not that important for master, as cockpit-tls usually does the TLS processing, which works around the issue. This is still relevant for the cockpit/ws container or custom setups, though.

**Comment by martinpitt** [comment]
The [unit test fails on ubuntu-1804](https://209.132.184.41:8493/logs/pull-13351-20200103-211313-019a21c3-cockpit-project-cockpit--ubuntu-1804/log.html) due to the unexpected message. ca-certificates is not present in the build chroot, just like with test-httpstream.

I pushed a fix.

**Comment by martinpitt** [comment]
> Where did you read about GTlsServerConnection being edge-triggered?

It's just an observation of its current behaviour. I didn't read about it explicitly, as neither https://developer.gnome.org/gio/stable/GPollableInputStream.html nor https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html explicitly document the behaviour -- so this patch simply removes the assumption that it has to be level triggered. Why do you think this is wrong? Is an extra call to `g_pollable_input_stream_read_nonblocking()` so expensive that we need to be concerned about it?

**Comment by martinpitt** [comment]
That said, of course I don't mind this being fixed in gio -- getting the fix into RHEL is a different story, of course. Do you think this would be appropriate for our stable branches, or is it completely wrong? (If so, I'm interested why). Thanks!

**Comment by martinpitt** [comment]
Curious: I just found https://gitlab.gnome.org/GNOME/glib/issues/805 from about a year ago (filed by @cgwalters). I wasn't aware of that before. @mvollmer already did pretty much the same fix in 2013 (!) in https://github.com/cockpit-project/cockpit/commit/de1622209e4b22b8c94e2b3dbf00b1da138f1058 , but that code was reorganized without a loop again in commit 2a0f62c850 , reintroducing the original problem.

File: src/common/cockpitwebserver.c [hunk 1]
```diff
@@ -1028,46 +1028,60 @@ on_request_input (GObject *pollable_input,
...
+      /* Once we receive data EOF is unexpected (until possible next request) */
+      request->eof_okay = FALSE;
     }
 
-  /* Once we receive data EOF is unexpected (until possible next request) */
-  request->eof_okay = FALSE;
-
-  return parse_and_process_request (request);
+  if (request->buffer->len > 0)
+    return parse_and_process_request (request);
```
Comment (stefwalter) [inline src/common/cockpitwebserver.c]: @martinpitt I think that this should be called within the loop for each successful read. It will both simplify the code, and also remain true to the original intent of how that function affects the request structure.

File: src/common/cockpitwebserver.c [hunk 2]
```diff
@@ -1028,46 +1028,60 @@ on_request_input (GObject *pollable_input,
   gsize length;
   gssize count;
 
-  length = request->buffer->len;
-  g_byte_array_set_size (request->buffer, length + 4096);
-
-  count = g_pollable_input_stream_read_nonblocking (input, request->buffer->data + length,
-                                                    4096, NULL, &error);
-  if (count < 0)
+  /* read all pending data */
```
Comment (stefwalter) [inline src/common/cockpitwebserver.c]: This should say:

/* 
 * HACK: https://gitlab.gnome.org/GNOME/glib-networking/issues/20
 * GLib incorrectly requires a looping in GPollableInputStream callback for TLS
 */

File: src/common/cockpitwebserver.c (hunk 1 above)
Comment (stefwalter) [inline src/common/cockpitwebserver.c]: I think that this should be called within the loop for each successful read. It will both simplify the code, and also remain true to the original intent of how that function affects the request structure.

Comment (martinpitt) [reply src/common/cockpitwebserver.c]: Interesting, I explicitly took it out of the loop, as it seems futile -- if there is more data, then the request definitively isn't complete, and `parse_and_process_request()` would just return False (hopefully), asking for more data.

Comment (stefwalter) [reply src/common/cockpitwebserver.c]: It seems that parse_and_process_request() could update request ... thereby affecting the loop. I think it's better just to call it iteratively.

Comment (stefwalter) [reply src/common/cockpitwebserver.c]: Oh hmmm. in addition there's another bug here. We should be calling parse_and_process_request until it returns FALSE. 

If indeed edge triggers, we not only need to read all input in a loop, but we need to process all data in the buffer in a loop.

Comment (martinpitt) [reply src/common/cockpitwebserver.c]: I'm afraid I still don't understand what you mean. `parse_and_process_request` sets `again` to true exactly when there is not enough data yet in the buffer, i. e. usually there's no terminating `\r\n` for the headers yet. This won't ever change when just calling it several times.

I initially thought that it will also be useless if we already know that there is more data in the buffer, as the request is clearly incomplete -- but apparently I misunderstand this, so I'm ok with moving it into the loop. But I'm not sure what you mean by "calling it until it returns FALSE"? If there's no more data in the buffer, and it returns FALSE, then we just leave the callback and wait for more data -- i. e. exactly how it behaves in current master. This PR doesn't change the request processing logic, just flushing the input buffer in the callback.

Comment (martinpitt) [reply src/common/cockpitwebserver.c]: Oh, do you mean that the `CockpitRequest` buffer could contain more than one request?

Comment (martinpitt) [reply src/common/cockpitwebserver.c]: I stared at this for a while, and I'm afraid I'm unable to see how `parse_and_process_request()` can sensibly be moved into the loop. If there is more data in the buffer and it returns FALSE, we have a contradiction -- it clearly processed the request too early and didn't notice that it was incomplete. Also, if it returns TRUE and we have no more data in the buffer, we must break the loop (and the current PR covers this case already).

The "good" cases are that we have no more data in the buffer and it returns FALSE (and the current PR covers that well), or that we have more data in the buffer and it returns TRUE (and the PR avoids this case, as that seems to be wasted time). So I'd appreciate your help with moving it into the loop.

Comment (martinpitt) [reply src/common/cockpitwebserver.c]: To clarify: The intention of this PR is to essentially change the g_pollable_input_stream_read_nonblocking() from a fixed buffer size to "as much as you have pending", not more, not less.

**Comment by martinpitt** [comment]
I realized this morning that we don't really need to do this in a loop. Request sizes are capped to 8 KiB anyway, so we can just bump the 4096 to 8192 and make this more robust by using the `cockpit_webserver_request_maximum` constant that we already have. I fixed that to be the true maximum (not half of it), and added a comment to make sure we don't break it again. We also have the integration test now, so we'll notice if we still manage to break it again. @allisonkarlitskaya , WDYT?

File: src/common/cockpitwebserver.c [hunk 3]
```diff
@@ -1029,10 +1029,15 @@ on_request_input (GObject *pollable_input,
...
+  /* Read all pending data; with some Gio versions and TLS this is not
+   * level-triggered, i. e. if there is more data available, the callback does
+   * not get called again; thus read up to our allowed maximum size. Plus one
+   * extra byte so that parse_and_process_request() correctly rejects requests
+   * that are > 8 KiB, instead of hanging. */
+  g_byte_array_set_size (request->buffer, length + cockpit_webserver_request_maximum + 1);
 
   count = g_pollable_input_stream_read_nonblocking (input, request->buffer->data + length,
-                                                    4096, NULL, &error);
+                                                    cockpit_webserver_request_maximum + 1, NULL, &error);
```
Comment (allisonkarlitskaya) [inline src/common/cockpitwebserver.c]: you need to subtract `length` here or you'll overrun the buffer

File: src/common/cockpitwebserver.c [hunk 4]
```diff
@@ -1029,10 +1029,15 @@ on_request_input (GObject *pollable_input,
   gssize count;
 
   length = request->buffer->len;
-  g_byte_array_set_size (request->buffer, length + 4096);
+  /* Read all pending data; with some Gio versions and TLS this is not
+   * level-triggered, i. e. if there is more data available, the callback does
```
Comment (allisonkarlitskaya) [inline src/common/cockpitwebserver.c]: please remove this language about edge- vs level- triggered.  The real issue is that we're not checking for pending data in the gnutls buffer.  But even then, I'd only mention that in the commit message and leave the code comment to just explain what we're trying to do here: read one more than the maximum request size to find out if we should hang up immediately or not.

File: src/common/cockpitwebserver.c [hunk 5]
```diff
@@ -883,7 +883,7 @@ parse_and_process_request (CockpitRequest *request)
   guint64 length;
 
   /* The hard input limit, we just terminate the connection */
-  if (request->buffer->len > cockpit_webserver_request_maximum * 2)
+  if (request->buffer->len > cockpit_webserver_request_maximum)
```
Comment (allisonkarlitskaya) [inline src/common/cockpitwebserver.c]: Probably it makes sense now to move this check after the read.

File: src/common/cockpitwebserver.c (hunk 3 above)
Comment (martinpitt) [reply src/common/cockpitwebserver.c]: No, the g_byte_array_set_size() above reserves length + cockpit_webserver_request_maximum + 1 bytes, so it should be fine. (The previous code did exactly the same, that didn't really change except for renaming the constant)

Comment (allisonkarlitskaya) [reply src/common/cockpitwebserver.c]: This is true.  But then the above code doesn't make sense: why should we be reserving space in the buffer to potentially read as much as 16k?  We should just fix the buffer at 8k+1, because it can't ever be smaller (given the request size) and should never be bigger (given the limit).

Comment (martinpitt) [reply src/common/cockpitwebserver.c]: Agreed, the current code is unnecessarily complicated. I really want to clean it up, but I prefer doing that in a follow-up (for backporting the minimally intrusive fix).

File: src/common/cockpitwebserver.c (hunk 5 above)
Comment (martinpitt) [reply src/common/cockpitwebserver.c]: Fully agreed, I'll clean this up in a follow-up.

File: src/common/cockpitwebserver.c (hunk 4 above)
Comment (martinpitt) [reply src/common/cockpitwebserver.c]: fixed

File: src/common/cockpitwebserver.c (hunk 3 above)
Comment (allisonkarlitskaya) [reply src/common/cockpitwebserver.c]: I don't think it would hurt to remove the `+ length` in the call to `g_byte_array_set_size()` and add a `- length` here.  I won't further block you if you want to push back on it, though.

File: src/common/cockpitwebserver.c [hunk 6]
```diff
@@ -40,7 +40,7 @@
 gboolean cockpit_webserver_want_certificate = FALSE;
 
 guint cockpit_webserver_request_timeout = 30;
-gsize cockpit_webserver_request_maximum = 4096;
+gsize cockpit_webserver_request_maximum = 8192;
```
Comment (allisonkarlitskaya) [inline src/common/cockpitwebserver.c]: Can we make this a `#define` please?  At the very least, `const`.

File: src/common/cockpitwebserver.c (hunk 3 above)
Comment (martinpitt) [reply src/common/cockpitwebserver.c]: @stefwalter just told me that this code is supposed to handle multiple pipelined requests, that's the reason for this structure. It's also why in the previous version we really should have called `process_request()` inside the loop -- it would look for an end of the request, process it, cut it out of the buffer, and let the next  reads just queue it back up. That's the reason behind checking for `cockpit_webserver_request_maximum * 2` and *not* having `- length` here.

So we actually have two different bugs here: (1) that issue with not calling the callback on pending data, and (2) 4 KiB is not enough for some legitimate SSO requests.

I'd like to fix this PR again to drop the ` * 2` change from above, fix the comments, and keep it to an absolute minimum. Then I'll rework #13372 to add a pipelining unit test and then see what still can be simplified.

File: src/common/cockpitwebserver.c (hunk 6 above)
Comment (martinpitt) [reply src/common/cockpitwebserver.c]: Right, I made it a const.

**Comment by martinpitt** [comment]
@allisonkarlitskaya : I now made this as small as possible. This still leaves open the corner case of "several large pipelined requests with TLS", but given that we don't even test pipelined requests, and it's apparently not something that ever bit us, I'll defer that to a follow-up.

**Comment by martinpitt** [comment]
I talked this through with @stefwalter: HTTP pipelining hasn't actually worked since 2014, it was most likely broken with commit 62bda3485c2. This should show two responses, but only shows the first:

    G_MESSAGES_DEBUG=all ./cockpit-ws -p 9999 --no-tls
    (printf 'GET /foo HTTP/1.1\r\nHost: test\r\n\r\nGET /cockpit/login HTTP/1.1\r\nHost: test\r\n\r\n'; sleep 2) | nc localhost 9999

Comparing that to a call to my server's Apache, which does deliver two responses:

    (printf 'GET / HTTP/1.1\r\nHost: localhost\r\n\r\nGET /tmp HTTP/1.1\r\nHost: localhost\r\n\r\n'; sleep 2) | nc piware.de 80

So we don't need to worry about the pipelining case for now. I'll give that some investigation and see how hard it is to salvage (I have an unit test ready for this).

File: test/verify/check-connection [hunk 7]
```diff
@@ -450,6 +450,13 @@ class TestConnection(MachineCase):
         output = m.execute('curl -s -S -k https://172.27.0.15:9000/ 2>&1 || true')
         self.assertIn('Connection refused', output)
 
+        # Large requests are processed correctly with plain HTTP
+        self.assertIn('A Custom Title', m.execute('''curl -s -S -H "Authorization: Negotiate $(printf '%0.7000i' 1)" http://localhost:9000/'''))
+
+        # Large requests are processed correctly with TLS
```
Comment (allisonkarlitskaya) [inline test/verify/check-connection]: In effect, this will be testing cockpit-tls, correct?

Comment (martinpitt) [reply test/verify/check-connection]: No, it tests cockpit-ws directly; the code above runs `cockpit_ws_executeable`. That's why it fails on 8-1-distropkg. We could add a test case through cockpit-tls as well of course, I can do that in a follow-up. I have another PR in the pipeline for adding a new unit test anyway.
